### ISSUE TITLE
SVG Image size problem

### ISSUE THREAD

**Issue by j-p-sequeira** [haraldk/TwelveMonkeys#518]
Hello, 

First, thank for the update! The ICO write support is very welcome :)

But, after updating from imageio-batik-3.4.2 to imageio-batik-3.5 I'm having problems resizing SVG images. Most likely, it is related to this update: [When rescaling SVGs use the ViewBox, if defined, for default sizes  ](https://github.com/haraldk/TwelveMonkeys/commit/24a0786d644c6fe434f79fbc153584fc86ab87c0)

Now to describe my problem: I do a bit of cheating to zoom SVG images. Every time the user changes the zoom in my viewer, I modify the XML data (width and height) and pass this new document to the reader to generate a new image. This way, the image is always zoomed with awesome SVG quality instead of using conventional raster zooming.

No need to say, in 3.4.2 all worked 100%. This is what I do:

```
        NodeList svgNodes = svgXML.getElementsByTagName("svg");
        if (svgNodes != null && svgNodes.getLength() > 0) {
            Node node = svgNodes.item(0);
            if (node instanceof Element) {
                String viewBox = ((Element) node).getAttribute("viewBox");
                if (viewBox.isEmpty()) {
                    // If no view box is in the original, there will be problems.
                    // Set a default one.
                    ((Element) node).setAttribute("viewBox", "0 0 " + widthPX + " " + heightPX);
                }

                ((Element) node).setAttribute("width", String.valueOf(widthPX));
                ((Element) node).setAttribute("height", String.valueOf(heightPX));
                success = true;
            }
        }
```

In 3.5, it all becomes messed up. I think 3.4.2 was actually correct and 3.5 is wrong, because if you open the images in browsers they will have the same size as in 3.4.2 (unless XML has no width/height specified in which the browser will have the size in relation to the available size inside the element but in my viewer I'll default it to 400px).

I think you'll be able to see the problem if you create copy of your SVG and try to double the size of the image. In 3.5, it is impossible because it ignores width/height, you'll need to halve the viewport but then the image gets cut.

I think the problem would be solved if you change the reader code to this (which makes much more sense to me):

```
            // get the 'width' and 'height' attributes of the SVG document
            Dimension2D docSize = ctx.getDocumentSize();
            if (docSize != null) {
                defaultWidth = (float) docSize.getWidth();
                defaultHeight = (float) docSize.getHeight();
            }
            else {
                SVGSVGElement rootElement = svgDoc.getRootElement();
                String viewBoxStr = rootElement.getAttributeNS(null, SVGConstants.SVG_VIEW_BOX_ATTRIBUTE);
                if (viewBoxStr.length() != 0) {
                    float[] rect = ViewBox.parseViewBoxAttribute(rootElement, viewBoxStr, null);
                    defaultWidth = rect[2];
                    defaultHeight = rect[3];
                }
                else {
                    defaultWidth = 200; //maybe 400 is better (not too small/big)
                    defaultHeight = 200; //maybe 400 is better (not too small/big)
                }
            }
```

**Comment by haraldk** [comment]
Thanks José!

Would you be able to make that into a (failing) test case, and suggest a fix as a PR, that would be great!

Otherwise, I'll try to look into the issue when I find the time.


Best regards,

-- 
Harald K

**Comment by j-p-sequeira** [comment]
I'm afraid I'm not able to do that since I lack the knowledge to do so.

But you can see that the code in the reader does not look very good, since you are always overwriting the width and height your retrieve from `Dimension2D docSize = ctx.getDocumentSize();` with the one in the viewbox if it exists (`viewBoxStr.length() != 0`). Might as well put `getDocumentSize()` in an else case when viewBoxStr has 0 length (unless you're expecting to caught an exception when looking for the viewbox and keep the previously retrieved sizes, but I don't think that's the case).

But bottom line, I'm not sure it is a good idea to use the viewbox size as the image size even when no width and height are present (much less when they are there).

**Comment by haraldk** [comment]
@Schmidor Do you have any input on this? I think you submitted the code to use `viewBox`, which seemed reasonable to me, but I'm no SVG expert. 😀

-- 
Harald K

**Comment by Schmidor** [comment]
I'm no svg expert either, but at least Photoshop generates svgs with only a viewbox definition. And there the viewbox ist definitively the intended base size.

**Comment by j-p-sequeira** [comment]
I'm just a newbie too, but I played a bit with Inkscape and the XML of SVG images. I'll try to explain what the viewbox is, since it caused me a lot of confusion before and I couldn't find any explanation on the matter. But it's because it's actually pretty simple once you realize it yourself.

First, an SVG image doesn't have a dimension concept. The fact that we are using vectors is that we can display the image in any size we want. However, when designing your company logo in SVG, you'll have different shapes and paths, all have positions and sizes in relation to each other.

Let me give you and example:
```
<svg xmlns="http://www.w3.org/2000/svg">
	<circle cx="0" cy="0" r="5" fill="red"/>
</svg>
```

We only have a circle with center at (0,0) and radius 5 (diameter 10). Now, these are not pixels. This is what's called SVG user space. If you save that code in notepad, name it with an SVG extension and open it in a browser, you might expect to see a 10px diameter circle, but you'll be in for a surprise.

What you see is a quarter of a circle (well, at least the 5px radius was as expected). But why?

If you play with inkscape (never tried photoshop) you'll notice a coordinate system. Those coordinates are the SVG user space. You can start drawing shapes and paths anywhere, even in negative coordinates. If you have several objects and you want to display the SVG, which is going to be the canvas size? One would expect (perhaps), it to be one that just tightly fits all objects, without margins, and that the software would do this for you. But you actually have to define it yourself. That's the viewbox.

In the example, to have it as expected, one has to define the view box. Viewbox is min-x, min-y, width and height. Width and height is easy: to tightly fit our circle, we need 10x10. The first two numbers are the min x and y. Since the circle is centered at point (0,0), we need -5,-5. So the viewbox we should set is this:
```
<svg viewBox="-5 -5 10 10" xmlns="http://www.w3.org/2000/svg">
	<circle cx="0" cy="0" r="5" fill="red"/>
</svg>
```

Now it should work as intended, right? Wrong! Check your browser for another surprise. Your circle tightly fits the available browser window. It didn't become and oval because SVG by default keeps the aspect ratio (you can use a flag to disable this behavior) and so, you get an image of the size of the browser window with transparent margin on the sides to keep the ratio.

Why did this happen? Because the browser is defaulting the size of the image to width=100% and height=100%. To have it as intended, you have to define the size of the image. The circle radius does not matter (it matters when there are more shapes, so we can have size relations between shapes).

For everything to works as one would expect, you need this:
```
<svg width="10" height="10" viewBox="-5 -5 10 10" xmlns="http://www.w3.org/2000/svg">
	<circle cx="0" cy="0" r="5" fill="red"/>
</svg>
```

Now, that width and height are actually finally in pixels. In fact, you can type "10px" instead, but px is the default unit.

Now, you can actually understand that using the viewBox="-5 -5 20 20" and the width="20" and height="20", you still actually get a 10px diameter circle with a 10px transparent margin at the right and at the bottom.

The beauty of SVG is that you define your viewbox and then rescale the whole image just by playing with the width and height.

In your reader now you use the viewbox to define the image size. If I want a bigger image, increasing the viewbox size will only increase the resulting image size by adding a transparent margin to the right and bottom. I actually have to half the view box (so then our shapes will look twice the size), but the the resulting image is small and everything will get clipped. By changing the viewbox size, I would need to update all shapes dimensions and coordinates to properly achieve the intended effect. That's very difficult and troublesome to do and I'd never get it right (CSS fonts would also needed to be scaled and etc...).

That is why you need to absolutely give priority to the width and height and not the viewbox. Now, what to do when (and this may happen if you do not define it in the software) the width and height are not set? You actually have 3 choices:

1) The right one is using 100% of available space. But that does not make sense at the twelvemonkeys' reader level. In my viewer, I could pass the available window size to the reader if you create a method to set available drawing size (though I'm not really interested in doing this).

2) Default to a size of your choice (I like this one better, I was actually injecting a 400x400 dimension, I can tell you why if you want).

3) Default to the size of the viewbox. This is not a bad choice and may be what everyone is expecting to see. But I don't like this for 3 reasons. First, viewbox is in SVG user space coordinates. Second, when designing your company SVG logo you can draw it (in SVG space) very big or very small. Very small quickly leads to fractional coordinates and results in loading tiny images. Very big, will result in getting a big image that when zooming can result in a very very big image. Third, this is not what the spec says.

Anyway, it does not matter which approach you use to me, you can go with 3) but you definitely must use the width and height if they are available, thus the code I suggested above. 

There is another problem: sizes can be specified in other units than px (you can use many others like percentage, inches or mm). If you want to design a 5mm bolt that is displayed on screen with 5mm you actually can. But this is going long already. Next, I'm going to create an example set of SVG images so that you can test your reader to see if the resulting images sizes are right, and I'll share my code to convert other SVG size units to pixels, since you may have a problem if ctx.getDocumentSize() is not doing it already for you (I'll have to run some tests, I think I remember converting the width and height to pixels before feeding the XML to your reader to fix this problem, but it was 2 years ago I need to refresh my memory).

**Comment by j-p-sequeira** [comment]
[svg_size_test.zip](https://github.com/haraldk/TwelveMonkeys/files/4133558/svg_size_test.zip)

Okay, attached is a folder with some SVG samples. In all of them, the orange logo should have 192x192px. You can check them in your browser (I'm using FF, but any browser except IE should do xD ). Even if you don't trust me (I mean, even I don't trust me!), please trust your browser.

In 3.4.2 I can achieve it. In 3.5, all of them will become 100px, because that was the original viewbox size (and to load the image in 192px you'll see how it is impossible just by messing with the header sizes, you'll have a major headache to achieve it).

And I'm indeed converting the width and height to pixels before passing the XML to the reader. So, even in 3.4.2 you may not see all images with 192px when other dimensions are used (mm/in...). If this is the case, I can help with the code to convert it to pixels, just say it if you need it ;)

**Comment by Schmidor** [comment]
The problem with your proposed change is, that Batik doesn't just return "null" if no width / height is set. Unfortunately `Dimension2D docSize = ctx.getDocumentSize();` defaults to 400 / 400 if missing and this then causes wrong sizes and scaling if no size is set in the svg. I think that was why if preferred the viewBox definition over getDocumentSize.

**Comment by j-p-sequeira** [comment]
> 
> 
> The problem with your proposed change is, that Batik doesn't just return "null" if no width / height is set. Unfortunately `Dimension2D docSize = ctx.getDocumentSize();` defaults to 400 / 400 if missing and this then causes wrong sizes and scaling if no size is set in the svg. I think that was why I preferred the viewBox definition over getDocumentSize.

I won't argue if the 400px default is any better or worse than using the viewbox size (both approaches are wrong`*`). The problem is that as of 3.5, twelvemonkeys is ignoring the `getDocumentSize `when it exists (the code is badly nested in the if-else). If there is a size it must be used. Otherwise, it's not just wrong. It's hell wrong!

But now that you've pointed it, I see that the problem may be that `getDocumentSize ` does not return null/exception when the size is not set. If it returns a Dimension2D with 400x400 we won't know if it was due to an error or if there really is a 400x400 set in the XML.... I'm not using Batik classes at all, so I have no experience on that. But if that's the case, the reader will have to look into the XML manually.

`*`As I said, when there is no size set, the image should expand to the available space. The reader only knows this available space if the application calling the reader somehow informs what that size is. Once set, that size should be used whenever the size set on the XML are in percentage (`size*percentage/100`) or there is no size set at all in the XML (that is, use 100% or `size*100/100 = size`). This is the proper way and how most devs will need it. Personally I don't need it, though. I'm already defining the size I want in the XML when there isn't one (though that is cheating). I could use my viewer window size (to be in spec, though I'm breaking spec anyway, I'll tell you why if you're curious).

`**` I will provide the code to get the size directly from XML, I'll make another post.

**Comment by j-p-sequeira** [comment]
Here goes the code.

Retrieve XML Size with plain Java (no Batik needed):
```
    /**
     * Gets the size of the SVG by reading it from the XML.
     *
     * @param svgXML - SVG XML document containing the vector information.
     * @return The pixel size. Never null. If the size is in percentage, invalid
     * or not present, -1 is returned.
     */
    public static Dimension getSVGPXSize(Document svgXML) {
        int widthPX = -1;
        int heightPX = -1;
        
        NodeList svgNodes = svgXML.getElementsByTagName("svg");
        if (svgNodes != null && svgNodes.getLength() > 0) {
            Node node = svgNodes.item(0);
            if (node instanceof Element) {
                String width = ((Element) node).getAttribute("width").trim();
                String height = ((Element) node).getAttribute("height").trim();
                width = width.isEmpty() ? "100%" : width;
                height = height.isEmpty() ? "100%" : height;

                try {
                    widthPX = ImageUtils.convertSVGSizeToPixels(width);
                } catch (AWTError | HeadlessException | NullPointerException | NumberFormatException ex) {
                    ImageUtils.LOG.log(Level.FINE, ex.getLocalizedMessage(), ex);
                    widthPX = -1; // Unecessary.
                }

                try {
                    heightPX = ImageUtils.convertSVGSizeToPixels(height);
                } catch (AWTError | HeadlessException | NullPointerException | NumberFormatException ex) {
                    ImageUtils.LOG.log(Level.FINE, ex.getLocalizedMessage(), ex);
                    heightPX = -1; // Unecessary.
                }
            }
        }

        return widthPX > 0 && heightPX > 0 ? new Dimension(widthPX, heightPX) : null;
    }
```

Translation to pixel (there are multiple problems here `*`). Can have spaces around unit:
```
    /**
     * Translates a SVG size to pixels.
     *
     * @param size - Size to translate (ex: "10in").
     * @return The translated size or exception on error.
     * @throws java.lang.NullPointerException
     * @throws java.lang.NumberFormatException
     * @throws java.awt.HeadlessException
     * @throws java.awt.AWTError
     */
    public static int convertSVGSizeToPixels(String size) throws NullPointerException, NumberFormatException, HeadlessException, AWTError {
        String cleanedSize = size.trim().toLowerCase();
        int sizeInPixel;

        if (cleanedSize.endsWith("px")) {
            cleanedSize = cleanedSize.substring(0, cleanedSize.length() - 2).trim();
            sizeInPixel = (int) (0.5 + Double.parseDouble(cleanedSize));
        } else if (cleanedSize.endsWith("%")) {
            cleanedSize = cleanedSize.substring(0, cleanedSize.length() - 1).trim();
            sizeInPixel = (int) (0.5 + (4 * Double.parseDouble(cleanedSize)));
        } else if (cleanedSize.endsWith("in")) {
            cleanedSize = cleanedSize.substring(0, cleanedSize.length() - 2).trim();
            sizeInPixel = (int) (0.5 + (Double.parseDouble(cleanedSize) * ApplicationUI.getScreenResolution()));
        } else if (cleanedSize.endsWith("mm")) {
            cleanedSize = cleanedSize.substring(0, cleanedSize.length() - 2).trim();
            sizeInPixel = (int) (0.5 + (Double.parseDouble(cleanedSize) * ApplicationUI.getScreenResolution()) / 25.4);
        } else if (cleanedSize.endsWith("cm")) {
            cleanedSize = cleanedSize.substring(0, cleanedSize.length() - 2).trim();
            sizeInPixel = (int) (0.5 + (Double.parseDouble(cleanedSize) * ApplicationUI.getScreenResolution()) / 2.54);
        } else if (cleanedSize.endsWith("pt")) {
            cleanedSize = cleanedSize.substring(0, cleanedSize.length() - 2).trim();
            sizeInPixel = (int) (0.5 + (Double.parseDouble(cleanedSize) * ApplicationUI.getScreenResolution()) / 72);
        } else if (cleanedSize.endsWith("pc")) {
            cleanedSize = cleanedSize.substring(0, cleanedSize.length() - 2).trim();
            sizeInPixel = (int) (0.5 + 12 * (Double.parseDouble(cleanedSize) * ApplicationUI.getScreenResolution()) / 72);
        } else if (cleanedSize.endsWith("em")) {
            cleanedSize = cleanedSize.substring(0, cleanedSize.length() - 2).trim();
            sizeInPixel = (int) (0.5 + (16 * Double.parseDouble(cleanedSize)) * 72 / ApplicationUI.getScreenResolution());
        } else if (cleanedSize.endsWith("ex")) {
            cleanedSize = cleanedSize.substring(0, cleanedSize.length() - 2).trim();
            sizeInPixel = (int) (0.5 + (25.4 * Double.parseDouble(cleanedSize)) / 16);
        } else {
            sizeInPixel = (int) (0.5 + Double.parseDouble(cleanedSize));
        }

        return sizeInPixel;
    }
```

Get screen resolution:
```
    /**
     * Gets the screen resolution DPI.
     *
     * @return The screen resolution DPI.
     */
    public static int getScreenResolution() {
        int screenResolution;

        try {
            screenResolution = Toolkit.getDefaultToolkit().getScreenResolution();
        } catch (HeadlessException | Error ex) {
            screenResolution = 72;
        }

        return screenResolution;
    }
```

`*` Problems are:
1- When % is used to set the size, I default to 400px. You can change this behavior to go look into the viewbox if you feel there like it.
2- You may be running twelvemonkey in an headless system (no display) so the screen resolution will fail and the pixel conversion will not make any sense.
3- There are actually more relative units and they are a mess to use. I attempt do deal with "em" and "ex" but that is not correct. And there are more units: [see here](https://www.w3schools.com/cssref/css_units.asp).
4-When conversion fails, null is returned. Again, default to whatever you like then.
5-Screen resolution defaults to 72dpi in case of failure. Maybe 96 is better or 600 (for printing)?

**Comment by j-p-sequeira** [comment]
Oh, by the way, to get the Document (and it supports SVGZ :) ):

```
    /**
     * Loads the XML document of the given SVG file.
     * 
     * @param svgFile - SVG file to load (*.svg or *.svgz)
     * @return The XML document of the given SVG file.
     */
    public Document loadSVG(File svgFile) {
        Document document = null;
        try {
            DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
            dbf.setNamespaceAware(false);
            dbf.setValidating(false);
            dbf.setExpandEntityReferences(false);
            dbf.setCoalescing(true);
            dbf.setIgnoringComments(false);
            dbf.setFeature("http://xml.org/sax/features/namespaces", false);
            dbf.setFeature("http://xml.org/sax/features/validation", false);
            dbf.setFeature("http://apache.org/xml/features/nonvalidating/load-dtd-grammar", false);
            dbf.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);

            boolean isSVGZ = FileUtils.getFileExtension(svgFile, false).equalsIgnoreCase("svgz");

            if (isSVGZ) {
                document = dbf.newDocumentBuilder().parse(new GZIPInputStream(new FileInputStream(svgFile)));
            } else {
                document = dbf.newDocumentBuilder().parse(new FileInputStream(svgFile));
            }
        } catch (ParserConfigurationException | SAXException | IOException ex) {
            FileManager.LOG.log(Level.WARNING, ApplicationInternationalization.getString("Error obtaining SVG XML information.")); // NOI18N
            FileManager.LOG.log(Level.FINE, ex.getLocalizedMessage(), ex);
        }
        
        return document;
    }
```

I don't think it is needed, but to check extension of a file:
```
    /**
     * Return the extension portion of the file's name.
     *
     * @param file - File from which to get the extension.
     * @param includeDot - Indicates if the returned string should include the
     * '.'.
     * @return The extension of the file or an empty string if there isn't any.
     */
    public static String getFileExtension(File file, boolean includeDot) {
        String ext = "";

        if (!file.isDirectory()) {
            int pointPosition = file.getName().lastIndexOf('.');

            if (pointPosition > 0) {
                // Not >=0 because if point is at the beginning, it is a hidden file and not an extension.
                ext = file.getName().substring(includeDot ? pointPosition : pointPosition + 1);
            }
        }
        return ext;
    }
```

**Comment by Schmidor** [comment]
Batik should already be able to parse those things. I've tried a fix, but now some tests in WMFImageReaderTest fail.
Could you try if this changes generate the expected results for you?

**Comment by Schmidor** [comment]
Ah, the build checks run through. Might be they depend on the system dpi :)

**Comment by j-p-sequeira** [comment]
I'm having problems building, can you send me the jar? Only the imageio-batik one will suffice.

Reading the code, it's starting to make much more sense now! 

I'm just not sure of what this `UnitProcessor.svgToUserSpace` does, and if it works as intended... And even if it does, what will happen backstage when a percentage is used?

And I was just wondering now, what happens when only the width or only the height is specified? The browser does not complain and behaves as expected. My code fails when using 3.4.2, as the dimension that is not specified falls back to 400px. Urgh, life is hard ( ; _ ; )

**Comment by Schmidor** [comment]
It uses the context in which it is rendered as reference. So for "em" probably the default font size and for percentages the view-area as reference size.

Do you know what the browser uses for the other dimension, if only width or only height is specified?
Currently it falls back to 200, as the 400 was from batik.

[twelvemonkeys-imageio-batik-3.6-SNAPSHOT.zip](https://github.com/haraldk/TwelveMonkeys/files/4134976/twelvemonkeys-imageio-batik-3.6-SNAPSHOT.zip)

**Comment by j-p-sequeira** [comment]
Thanks! I'll test with this new JAR and will report back very soon.

The browser calculates the missing dimension to keep aspect ratio. It must be from the viewbox ratio:

`missing  width = vb_width * height / vb_height`

or 

`missing height = vb_height * width / vb_width`

But I don't know what happens when preserveAspectRatio="none" is used or if no viewbox is defined (defaults to 0 0 300? 150?).

**Comment by j-p-sequeira** [comment]
I tested with the new JAR and it's perfect! My viewer displays the images sized as intended! Thanks a lot!!! Zoom is working again :)

Of course, I did not test with all acute rare cases mentioned before. I'll also have to fix the missing dimension problem in my code (even if you fix it in the reader, I think my zooming code will trigger it back), but that will be for another day.

**Comment by Schmidor** [comment]
I've added some of those corner cases with fallbacks. If we don't get sample images that show otherwise it should do for now :)

**Comment by j-p-sequeira** [comment]
Well, indeed, looks good to me as it is. Awesome work!

**Comment by haraldk** [comment]
I belive everything we need to know is here:

https://www.w3.org/TR/SVG/coords.html

Unfortunately, the `preserveAspectRatio` attribute has 20 different combos (10x meet + 10x slice)...

But perhaps we should only worry about the initial viewport size and not have to worry about all that. 

PS: @j-p-sequeira You are aware that the `SVGImageReader` has a special `SVGImageReadParam` that allows specifying a size via the [`setSourceRenderSize(Dimension)`](https://docs.oracle.com/javase/7/docs/api/javax/imageio/ImageReadParam.html#setSourceRenderSize(java.awt.Dimension)) method? I have a feeling this should be a much cleaner solution than messing with the XML directly...

-- 
Harald K

**Comment by haraldk** [comment]
Fixed via #522 .
